CORE-8962 flow mapper absorbs messages when in state closing

emilybowe opened the pull request:
**Description:**
When a session has completed the Flow Mapper holds onto its state and sets it to CLOSING. It now ignores any incoming messages. This means if an ACK message is lost that one party cannot complete its close and will log the session terminated in error while the other party will record it completed successfully. This change will send automatic ACKS in that situation to prevent this. This PR also introduces a new state, ERROR. 

**Links:**
Ticket -> [CORE-8962: Flow Mapper swallows all messages it receives when it is in state CLOSING](https://r3-cev.atlassian.net/browse/CORE-8962)

Api pr -> [CORE-8962: api change for flow mapper absorbs messages (https://github.com/corda/corda-api/pull/1020)

//update api number before merge

emilybowe committed:
rm unused exception

emilybowe committed:
rm unused exception

emilybowe committed:
adding new class for SessionError, changing processOtherSessionEvents in SessionEventExecutor

emilybowe committed:
Merge remote-tracking branch 'origin/release/os/5.0' into CORE-8962_Flow_Mapper_absorbs_messages_when_in_state_CLOSING

emilybowe committed:
Merge remote-tracking branch 'origin/CORE-8962_Flow_Mapper_absorbs_messages_when_in_state_CLOSING' into CORE-8962_Flow_Mapper_absorbs_messages_when_in_state_CLOSING

emilybowe committed:
adding to SessionErrorExecutor

emilybowe committed:
add notes for design

emilybowe committed:
generate ack

emilybowe committed:
add to notes

emilybowe committed:
adding to processOtherSessionEvents

emilybowe committed:
more work on SessionErrorExecutor

emilybowe committed:
adding new SessionErrorExecutor to FlowMapperExecFactoryImpl

emilybowe committed:
rm isValid and log msg

emilybowe committed:
send error back to flow if received in OPEN state

emilybowe committed:
fix parameters for FlowMapperEventExecutorFacImpl

emilybowe committed:
handle null in exhaustive when clauses

emilybowe committed:
rm duplicate test, update CLOSING test as now events are processed

emilybowe committed:
Merge remote-tracking branch 'origin/release/os/5.0' into CORE-8962_Flow_Mapper_absorbs_messages_when_in_state_CLOSING

# Conflicts:
#	gradle.properties

emilybowe committed:
add gradle tag for api and fix null unit test

emilybowe committed:
fix tests, add unit test for error class

emilybowe committed:
add osgi debugging in test.bndrun

emilybowe committed:
integration tests

emilybowe committed:
adding tests

emilybowe committed:
adding to tests

emilybowe committed:
Merge branch 'release/os/5.0' into CORE-8962_Flow_Mapper_absorbs_messages_when_in_state_CLOSING

emilybowe committed:
bummp version

emilybowe committed:
handle FlowMapperState with null status

emilybowe committed:
trigger ERROR state when we receive a SessionError

LWogan commented on components/flow/flow-mapper-service/src/main/kotlin/net/corda/session/mapper/service/executor/FlowMapperMessageProcessor.kt:
```diff
@@ -34,40 +31,28 @@ class FlowMapperMessageProcessor(
-            logger.debug { "Ignoring event (isExpiredSessionEvent: ${isExpiredSessionEvent(value)}, " +
-                    "isValidState: ${isValidState(state, value)})" }
+            logger.debug {
+                errorMsg.format("ignored", isExpiredSessionEvent(value), state)
```
I think isExpiredSessionEvent(value) in this case will always be true so can avoid double processing by either just hardcoding true, or assigning the result of the method to a val and using that in the conditional check and in the else.

mbrkic-r3 replied:
Message expects event as the second argument, not expiration status

mbrkic-r3 replied:
Changed code to send `event` as the second argument

emilybowe committed:
bump version again

emilybowe committed:
Merge branch 'release/os/5.0' into CORE-8962_Flow_Mapper_absorbs_messages_when_in_state_CLOSING

LWogan commented on libs/flows/flow-mapper-impl/src/main/kotlin/net/corda/flow/mapper/impl/FlowMapperEventExecutorFactoryImpl.kt:
```diff
@@ -43,8 +45,26 @@ class FlowMapperEventExecutorFactoryImpl @Activate constructor(
                 val eventPayload = sessionEvent.payload
                 if (eventPayload is SessionInit) {
                     SessionInitExecutor(eventKey, sessionEvent, eventPayload, state, sessionEventSerializer, flowConfig)
```
We should change the if ... else if ... else to a `when`. and maybe extract that additional when to another function to avoid nested whens which would be an eye sore

LWogan commented on libs/flows/flow-mapper-impl/src/main/kotlin/net/corda/flow/mapper/impl/executor/SessionErrorExecutor.kt:
```diff
@@ -0,0 +1,103 @@
+            }
+            FlowMapperStateType.OPEN -> {
+                log.warn(errorMsg + "Forwarding event.")
+                flowMapperState.status = FlowMapperStateType.ERROR
```
Long winded comment incoming. In short I think this change is OK :D

I can see why you'd change the state from OPEN to ERROR. We have received an ERROR. This does raise an interesting point though

At the moment the only thing that sets the Flow Mapper state is the flow engine that created it. Up to now this was done by creating a new session via the first SessionInit event going out or coming in(sets to OPEN), and ending a session/flow and sending the Mapper a ScheduleCleanup event (sets to CLOSING),  Now we are adding an additional case where we send out a SessionError and this triggers the mapper state to go to ERROR.

Additionally this logic means if a counterparty sends an ERROR to us, it will also change the mapper status to be ERROR. I don't think this change would cause problems right now. But we might in the future change the flow engine to examine the session error and let it decide if it is in fact an error or not. If the flow engine decided to say "hey this isn't an error I care about, I'm going to ignore it" but the mapper has closed the gate and stopped letting messages through we could be in trouble. The flow engine could be willing to accept more messages but the mapper has already been set to ERROR. However, that use case does not exist yet. All errors are legit errors which is why this logic is fine.

If we changed this to ignore SessionErrors being receive from the other side it would mean wed have to add logic to the flow engine to send a message to the mapper after receiving an error to tell it to set to ERROR. We do something similar right now in the global post processor to schedule mapper cleanup. But yeah, for now, I think always setting the mapper state to ERROR works.

LWogan commented on libs/flows/flow-mapper-impl/src/main/kotlin/net/corda/flow/mapper/impl/executor/SessionErrorExecutor.kt:
```diff
@@ -0,0 +1,103 @@
+                    FlowMapperResult(
+                        flowMapperState, listOf(
+                            Record(
+                                Schemas.P2P.P2P_OUT_TOPIC, sessionId, appMessageFactory(
```
I think we have this logic to generate a record for p2p in two places, so it would be good to extract it out to a common method shared between the 2 classes.

mbrkic-r3 replied:
Refactored duplicate code into function "createP2PRecord"

LWogan reviewed (commented):
I've done a first pass. And I think it looks correct. I'll need to go back over it again though in more depth. 

I do think we're missing a small piece though. The flow mapper only holds onto states for a certain period of time after they error or close. We have logic in the FlowGlobalPostProcessor.getFlowMapperSessionCleanupEvents to send a ScheduleCleanup message to the mapper when a session errors or closes. This message also sets the mapper state to be CLOSING.  

So I think this scenario will happen when we receive a SessionError from a counterparty:
- mapper receives error, sets status to be ERROR
- mapper sends error to flow engine.
- flow engine changes it session state status to be ERROR
- FlowGlobalPostProcessor sees a session state with status ERROR
- FlowGlobalPostProcessor sends a ScheduleCleanup() to mapper
- mapper receives ScheduleCleanup
- mapper sets status to be CLOSING and schedules cleanup
- counterparty sends a normal session message, and mapper responds with an ACK.

 I think you need to make a change to make sure that the state stays in ERROR when the mapper receives a ScheduleCleanup message (if it is already in state ERROR)

emilybowe committed:
change FlowMapperMessageProcessor unit tests and add more

emilybowe committed:
add state check to integration tests and check we change state to ERROR

emilybowe committed:
replace nested if with when

emilybowe committed:
change topic so we respond to CLOSING request from counterparty and send a P2P msg outbound.

emilybowe committed:
fix unit test

emilybowe committed:
add state to unit tests

emilybowe committed:
add state to tests

LWogan commented on libs/flows/flow-mapper-impl/src/main/kotlin/net/corda/flow/mapper/impl/FlowMapperEventExecutorFactoryImpl.kt:
```diff
@@ -38,21 +40,49 @@ class FlowMapperEventExecutorFactoryImpl @Activate constructor(
+                            flowConfig
+                        )
+                    }
+                    is SessionError -> {
```
I think we're missing a unit test for SessionErrorExecutor in FlowMapperEventExecutorFactoryImplTest

mbrkic-r3 replied:
Added missing test

mbrkic-r3 committed:
CORE-8962 Updated API version and small fixed based on code review:
- Fixed parameter value for "errorMsg" in FlowMapperMessageProcessor
- Refactored duplicate code into function "createP2PRecord"
- Updating FlowMapper state in ScheduleCleanupEventExecutor to CLOSING only if it was not ERROR
- Added test to FlowMapperEventExecutorFactoryImplTest for SessionErrorExecutor

mbrkic-r3 commented:
> I've done a first pass. And I think it looks correct. I'll need to go back over it again though in more depth.
> 
> I do think we're missing a small piece though. The flow mapper only holds onto states for a certain period of time after they error or close. We have logic in the FlowGlobalPostProcessor.getFlowMapperSessionCleanupEvents to send a ScheduleCleanup message to the mapper when a session errors or closes. This message also sets the mapper state to be CLOSING.
> 
> So I think this scenario will happen when we receive a SessionError from a counterparty:
> 
> * mapper receives error, sets status to be ERROR
> * mapper sends error to flow engine.
> * flow engine changes it session state status to be ERROR
> * FlowGlobalPostProcessor sees a session state with status ERROR
> * FlowGlobalPostProcessor sends a ScheduleCleanup() to mapper
> * mapper receives ScheduleCleanup
> * mapper sets status to be CLOSING and schedules cleanup
> * counterparty sends a normal session message, and mapper responds with an ACK.
> 
> I think you need to make a change to make sure that the state stays in ERROR when the mapper receives a ScheduleCleanup message (if it is already in state ERROR)

Updated `FlowMapper` state in `ScheduleCleanupEventExecutor` to `CLOSING` only if it was not `ERROR`

mbrkic-r3 commented on libs/flows/flow-mapper-impl/src/main/kotlin/net/corda/flow/mapper/impl/executor/SessionErrorExecutor.kt:
```diff
@@ -0,0 +1,94 @@
+                                sessionEventSerializer,
+                                appMessageFactory,
+                                flowConfig,
+                                0
```
Would `sessionEvent.receivedSequenceNum` have value 0 here so we don't need to send it?

LWogan replied:
not a big deal. but the outbound message is one which originated from within the flow engine. so we can set  receivedSequenceNum as sessionEvent.receivedSequenceNum here

mbrkic-r3 replied:
modified

mbrkic-r3 commented on libs/flows/flow-mapper-impl/src/main/kotlin/net/corda/flow/mapper/impl/executor/SessionEventExecutor.kt:
```diff
@@ -38,54 +39,91 @@ class SessionEventExecutor(
+                        sessionEventSerializer,
+                        appMessageFactory,
+                        flowConfig,
+                        0
```
Would `sessionEvent.receivedSequenceNum` have value 0 here so we don't need to send it?

LWogan replied:
I think 0 is ok. This scenario is for when a session is expired so we cant set it sensibly. The SessionError will trigger the error logic on the other side anyway so the field wont be used.

mbrkic-r3 committed:
CORE-8962 Added missing param to documentation of "createP2PRecord"

mbrkic-r3 committed:
Merge remote-tracking branch 'origin/release/os/5.0' into CORE-8962_Flow_Mapper_absorbs_messages_when_in_state_CLOSING

# Conflicts:
#	gradle.properties

LWogan commented on libs/flows/flow-mapper-impl/src/main/kotlin/net/corda/flow/mapper/impl/executor/SessionEventExecutor.kt:
```diff
@@ -38,54 +39,91 @@ class SessionEventExecutor(
+                } else {
+                    val outputRecord =
+                        createP2PRecord(
+                            sessionEvent,
```
I think this needs to be changed. The SessionEvent is used to carry the ack info on the receivedSequenceNum and outOfOrderSequenceNums fields. If we copy the event they sent us back to them, it will not give valid ack info. We need to set outOfOrderSequenceNums = emptylist and we need to set sessionEvent.receivedSequenceNum = sessionEvent.sequenceNum

LWogan replied:
we should also set sessionEvent.sequenceNum = null as messages with SessionAck as a payload are not allowed have sequence numbers

mbrkic-r3 replied:
`emptyList` and for `outOfOrderSequenceNums` and `null` for `sequenceNum` are already used in `createP2PRecord`

LWogan replied:
ah i see. ok then you just need to update how receivedSequenceNumber is set

mbrkic-r3 replied:
Set `sessionEvent.sequenceNum` as default value for `receivedSequenceNumber` in `createP2PRecord`

LWogan commented on libs/flows/flow-mapper-impl/src/main/kotlin/net/corda/flow/mapper/impl/executor/FlowMapperHelper.kt:
```diff
@@ -76,3 +77,41 @@ fun generateAppMessage(
+                null,
+                sessionEvent.initiatingIdentity,
+                sessionEvent.initiatedIdentity,
+                receivedSequenceNumber,
```
should be set to sessionEvent.sequenceNumber

mbrkic-r3 replied:
Should be fixed by the change mentioned in the comment above

mbrkic-r3 committed:
CORE-8962 Setting sessionEvent.receivedSequenceNum = sessionEvent.sequenceNum when sending ACK for CLOSING state

mbrkic-r3 committed:
CORE-8962 Set sessionEvent.sequenceNum as default value for receivedSequenceNumber in createP2PRecord

LWogan reviewed (dismissed):
These changes look good to me!

mbrkic-r3 committed:
CORE-8962 Sending ACK when CLOSING only for SessionClose

mbrkic-r3 committed:
CORE-8962 Fixed failing unit test "Session event received with CLOSING state"

LWogan reviewed (dismissed):
These changes look good to me. This handles the infinite ack problem we uncovered late on.

mnesbit reviewed (dismissed):
OK

mbrkic-r3 committed:
CORE-8962 API version set to 5.0.0.752-beta+

thiagoviana reviewed (approved):
Re-Approving it.
